### PR TITLE
Allow importing of memory.

### PR DESCRIPTION
**Pull request by koponen-styra**
This is still missing unit tests but wanted to get feedback first whether this is the right direction to add support for memory imports.

### CODE REVIEW

**Comment by Hywan** [comment]
For what I read, the patch is in good tracks. We want to abandon the old import API and focus on the `ImportObject` API. That's the correct way to go.

Thanks for working on it!

**Comment by koponen-styra** [comment]
Do you mean instead of adding AppendFunction to Imports, have another constructor for ImportObject to construct a memory import? Sorry, just lacking still a lot of context here. Concrete suggestions are much appreciated.

**Comment by Hywan** [comment]
The idea is to use https://wasmerio.github.io/wasmer/c/runtime-c-api/wasmer_8h.html#a9b254ed5edbb735de3e9c2a83ddf6040 by passing https://wasmerio.github.io/wasmer/c/runtime-c-api/structwasmer__import__t.html with `tag` set to `WASM_MEMORY`. And so you need to update https://github.com/wasmerio/go-ext-wasm/blob/master/wasmer/import.go#L67 to support any kind of imports, not just `WASM_FUNCTION`.

**Comment by koponen-styra** [comment]
The PR actually already changes the ImportObject's Extend to handle wasmer_import_t with memory tag, by improving getCWasmerImport to handle the memory imports as well. So perhaps after all, the main thing missing are tests.

**Comment by Hywan** [comment]
I'll review your PR as soon as possible.

File: wasmer/test/imports.go [hunk 1]
```diff
@@ -292,3 +292,38 @@ func testWasiImportObject(t *testing.T) {
...
+	defer memory.Close()
+
+	imports, err = imports.AppendMemory("memory", memory)
+	assert.NoError(t, err)
+
+	importObject := wasm.NewImportObject()
+	err = importObject.Extend(*imports)
+	assert.NoError(t, err)
+
+	instance, err := module.InstantiateWithImportObject(importObject)
```
Comment (Hywan) [inline wasmer/test/imports.go]: I'm surprised it works. See my review comment. The test doesn't actually test the memory in this case.

File: wasmer/test/testdata/log_import_memory.wat [hunk 2]
```diff
@@ -0,0 +1,15 @@
+(module
+  (type (;0;) (func (param i32 i32)))
+  (type (;1;) (func))
+  (import "env" "log_message" (func (;0;) (type 0)))
+  (import "env" "memory" (memory (;0;) 17))
+  (func (;1;) (type 1)
+    i32.const 1048576
```
Comment (Hywan) [inline wasmer/test/testdata/log_import_memory.wat]: We don't read from the memory here, but from the statically allocated data, i.e. from the Data Section.

**Comment by Hywan** [comment]
I've opened https://github.com/koponen-styra/go-ext-wasm/pull/1 after my reviews. That's a small set of improvements.

**Comment by Hywan** [comment]
To be clear: Apart from the test that doesn't test the new feature, the code looks really great, good job!

File: wasmer/memory.go [hunk 3]
```diff
@@ -86,3 +108,8 @@ func (memory *Memory) Grow(numberOfPages uint32) error {
 
 	return nil
 }
+
+// Close closes/frees memory allocated at the NewMemory at time.
+func (memory *Memory) Close() {
+	cWasmerMemoryDestroy(memory.memory)
```
[the remark this answers is not in the document]
Comment (koponen-styra) [reply wasmer/memory.go]: Makes sense.

File: wasmer/test/testdata/log_import_memory.wat (hunk 2 above)
Comment (koponen-styra) [reply wasmer/test/testdata/log_import_memory.wat]: Ouch. Interesting it works... I'll fix it to read from memory.

File: wasmer/memory.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (koponen-styra) [reply wasmer/memory.go]: Resolving this since it's a part of the other PR.

**Comment by koponen-styra** [comment]
Let me know if you prefer me to pull in your commits from your PR on top of this one or whether you prefer the other way around.

**Comment by koponen-styra** [comment]
Fixed the test and made it minimal. In addition, noticed and fixed a minor issue with importObject.Imports: it considered modules with no function imports as invalid.

**Comment by syrusakbary** [comment]
@koponen-styra if you can pull @Hywan commits into this PR that would be great! It will accelerate a bit the process :)

File: wasmer/import.go [hunk 4]
```diff
@@ -35,11 +35,6 @@ func NewImportObject() *ImportObject {
 // Imports returns `*Imports` for a given `ImportObject`
 func (importObject *ImportObject) Imports() (*Imports, error) {
 	imports := cWasmerImportObjectGetFunctions(importObject.inner)
-
-	if imports == nil {
```
Comment (Hywan) [inline wasmer/import.go]: This is dangerous, but we need to fix this anyway.

**Pull request merged**